File: perl/ScalrAPIExample.pl
#!/usr/bin/perl

use strict;

use POSIX qw(strftime);
use MIME::Base64 qw(encode_base64);
use Digest::SHA qw(hmac_sha256);   
use LWP::UserAgent;                

# Constants
use constant {
        SCALR_API_KEY    => "xxxxxxxxxxxxxxx",
        SCALR_SECRET_KEY => "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
        API_URL          => "https://api.scalr.net/",
        API_VERSION      => "2.3.0",
        API_AUTH_VERSION => 3,
};

my $TZ = strftime("%z", localtime);
$TZ =~ s/(\d{2})(\d{2})/$1:$2/;    

# ISO8601
my $Timestamp = strftime("%Y-%m-%dT%H:%M:%S", localtime) . $TZ;

# API call parameters
my $params_ref = {
        Action => 'FarmsList',
        Version => API_VERSION,
        Timestamp => $Timestamp,
        KeyID => SCALR_API_KEY,
        AuthVersion => API_AUTH_VERSION,
};

my $string_to_sign;

$string_to_sign = $params_ref->{'Action'} . ':' . $params_ref->{'KeyID'} . ':' . $params_ref->{'Timestamp'};

$params_ref->{'Signature'} = encode_base64( hmac_sha256($string_to_sign, SCALR_SECRET_KEY) );

$params_ref->{'Signature'} =~ s/\n//g;

my $ua = LWP::UserAgent->new;

my $response = $ua->post(API_URL, $params_ref);

if ($response->is_success) {
        print $response->content;
}
else {
        print $response->as_string;
}

File: php/ScalrAPIExample.php
<?php

define('SCALR_API_KEY', 'xxxxxxxxxxxxxxx');
define('SCALR_SECRET_KEY', 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx');
define('API_URL', 'https://api.scalr.net/');
define('API_VERSION', '2.3.0');
define('API_AUTH_VERSION', '3');
define('TIMEZONE', 'UTC');

date_default_timezone_set(TIMEZONE);

// Build query arguments list
$params = array(
    'Action'     => 'FarmsList',
    'Version'    => API_VERSION,
    'Timestamp'  => date("Y-m-d H:i:s"),
    'KeyID'      => SCALR_API_KEY,
    'AuthVersion'=> API_AUTH_VERSION
);

// Generate string for sign
$string_to_sign = $params['Action'] . ":" . $params['KeyID'] . ":" . $params['Timestamp'];
$params['Signature'] = base64_encode(hash_hmac('SHA256', $string_to_sign, SCALR_SECRET_KEY, 1));

//GET request
// Build query
$query = http_build_query($params);

// Execute query
$c = file_get_contents(API_URL."?{$query}");
print_r($c);
